src/language/patterns.md

```diff
@@ -0,0 +1,334 @@
+on, or continue execution if the object doesn't match. Case patterns can be any
+kind of pattern:
+
+- [**Constant patterns**][constant]: the object's value has to be equivalent to match.
+  ```dart
+  switch (number) {
+    case 1: ...          // Matches if number == 1  
+  }
+  ```
+- [**List**][list], [**map**][map], or [**record**][record] **patterns**: the elements are destructured if matched.
+  ```dart
+  switch (obj) {
+    case [a, b]: ...     // Matches if obj is a list with two elements  
+  }
+  ```  
+- [**Variable patterns**][variable]: the matched value binds to a new variable.
+  ```dart
+  switch ((1, 2)) {
+    case (var a, var b): ...    // Matches if 1 and 2 can be assigned to a and b
+  }
+  ```  
+- [**Logical**][logical] or [**relational**][relational] **patterns**: the object matches if the comparison to
+a constant returns `true`.
+  ```dart
+  switch (code) {
+    case >= first && <= last: ...    
+    // Matches if the value of code is between the constant values of first and last
+  }
+  ```  
+- [**Null-check**][check] or [**null-assert**][assert] **patterns**: the match is first on whether the object is not null, then on the value.
+  ```dart
+  switch (maybe) {
+    case var s?: ... 
+    // Matches if maybe is not null, and then against the inner pattern "var s"
+  }
+  ```  
+- [**Object patterns**][object]: the match exposes getters on the object type.
+  ```dart
+  switch (Shape shape) {
+    case Rect(width: var w, height: var h): ...
+    // Matches if shape is of type Rect, and then against the properties of Rect
+  }
+  ```  
+
+Values destructured by a pattern in a case become local variables,
+only available in the body of that case. 
+
+### Object destructuring
+
+[Object patterns][object] match against named object types, allowing
+you to destructure their data using the getters the object's class already exposes.
+
+For example, you can use object destructuring in a [for-in loop][for] to iterate-over
+and destructure the [`MapEntry`][] objects that a `<Map>.entries` call
+returns:
+
+```dart
+Map<String, int> hist = // ...
+
+for (var MapEntry(key: key, value: count) in hist.entries) {
+  print('$key occurred $count times');
+}
+```
+
+The object pattern checks that `hist.entries` has the named type `MapEntry`,
+and then recurses into the named field subpatterns `key` and `value`.
+It calls the `key` getter and `value` getter on the `MapEntry` in each iteration,
+and binds the results to local variables `key` and `count`, respectively.
+
+Binding the result of a getter call to a variable of the same name is a common
+use case, so object patterns can also infer the getter name from the
+[variable subpattern][variable]. This allows you to simplify the variable pattern
+from something redundant like `key: key` to just `:key`:
+
+```dart
+for (var MapEntry(:key, value: count) in hist.entries) { 
+  print('$key occurred $count times');
+}
+```
```

Comment:
This section feels a little tacked on because it's more or less just a description of how object patterns work in general, so it overlaps the previous object pattern bullet point and the section on object patterns in the reference. Maybe expand/merge this into section that talks about using patterns in for/for-in loops (which isn't otherwise shown on this doc) and make this an example under that section?